lib/core/src/Cardano/Pool/DB.hs

```diff
@@ -166,6 +188,87 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
         -- For a Sqlite DB, this would be "run a query inside a transaction".
     }
 
+-- | Given the /latest/ registration and retirement certificates for a pool,
+--   determine the pool's current registration status based on the relative
+--   order in which the certificates were published.
+--
+-- If two certificates are supplied, then:
+--
+--   * the certificates must be from the same pool.
+--   * the publication times must be non-equal.
+--
+-- This function determines order of precedence according to the "pool
+-- inference rule", as described in "A Formal Specification of the Cardano
+-- Ledger":
+--
+-- https://hydra.iohk.io/build/3202141/download/1/ledger-spec.pdf
+--
+determinePoolRegistrationStatus
+    :: Ord certificatePublicationTime
+    => Maybe (certificatePublicationTime, PoolRegistrationCertificate)
+    -> Maybe (certificatePublicationTime, PoolRetirementCertificate)
+    -> PoolRegistrationStatus
+determinePoolRegistrationStatus mReg mRet = case (mReg, mRet) of
+    (Nothing, _) ->
+        PoolNotRegistered
+    (Just (_, regCert), Nothing) ->
+        PoolRegistered regCert
+    (Just (regTime, regCert), Just (retTime, retCert))
+        | regPoolId /= retPoolId ->
+            -- Comparing certificates from different pools is a programming
+            -- error.
+            differentPoolsError
+        | regTime > retTime ->
+            -- A re-registration always /supercedes/ a prior retirement.
+            PoolRegistered regCert
+        | regTime < retTime ->
+            -- A retirement always /augments/ the latest known registration.
+            PoolRegisteredAndRetired regCert retCert
+        | otherwise ->
+            -- If a registration certificate and a retirement certificate
+            -- for the same pool appear to have been published at exactly
+            -- the same time, this indicates a programming error.
+            timeCollisionError
```

Comment:
Same here - also the error messages should mention "programming error", so that if it ever occurs - which it won't - the user knows that it's not their fault.

Reply:
> Same here - also the error messages should mention "programming error", so that if it ever occurs - which it won't - the user knows that it's not their fault.

> You may as well remove the comment and inline the differentPoolsError string -- it's self-documenting.

Thanks! I've removed the comment, and revised the error string to make it clear that it's a programming error.